charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,281 @@
 
-    def _check_leader(self):
+    def _check_leader(self) -> None:
+        """Check whether a unit is the leader."""
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
-    def _check_image_details(self):
+    def _check_container_connection(self) -> None:
+        """Check if connection can be made with container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
+
+    def _check_status(self) -> None:
+        """Check status of workload and set status accordingly."""
+        self._check_leader()
+        try:
+            container = self.unit.get_container(self._container_name)
+        except ModelError as error:
+            raise GenericCharmRuntimeError(
+                f"Failed to fetch container {self._container_name}"
+            ) from error
         try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            raise CheckFailed(f"{e.status.message}", e.status_type)
-        return image_details
+            check = container.get_check(PROBE_NAME)
+        except ModelError as error:
+            raise GenericCharmRuntimeError(
+                f"Failed to run health check on workload container: {PROBE_NAME} not found"
+            ) from error
+
+        if check.status != CheckStatus.UP:
+            self.logger.error(
+                f"Container {self._container_name} failed health check {PROBE_NAME}. The"
+                " Tensorboard Controller service will be restarted."
+            )
+            self.unit.status = MaintenanceStatus(
+                "Workload failed health check, restarting controller service"
+            )
+        else:
+            self.unit.status = ActiveStatus()
+
+    def _on_install(self, _) -> None:
+        """Handle install event."""
+        # deploy K8s resources to speed up deployment
+        self._apply_k8s_resources()
+
+    def _on_upgrade(self, _) -> None:
+        """Handle upgrade event."""
+        # force conflict resolution in K8s resources update
+        self._on_event(_, force_conflicts=True)
+
+    def _on_remove(self, _) -> None:
+        """Handle remove event."""
+        delete_errors = []
+        self.unit.status = MaintenanceStatus("Removing K8s resources")
+        for handler in {self.k8s_resource_handler, self.crd_resource_handler}:
+            try:
+                handler.delete()
+            except ApiError as error:
+                delete_errors.append(error)
+
+        if delete_errors:
+            raise GenericCharmRuntimeError("Removing K8s resources failed", delete_errors)
+
+    def _on_update_status(self, _) -> None:
+        """Handle update status event."""
+        self._on_event(_)
```

Comment:
Carried this pattern over from the Jupyter Controller charm. If I'm being honest, I don't know what the rationale here is. I'd guess it's about running the main reconciliation loop before attempting to set the status. This is definitely something we should strive to perform uniformly throughout our charms.

Reply:
`_on_event()` does only replanning in addition to applying K8S resources, that's pretty much it for this charm.
Why `_on_event()` is called in `update_status`? It is to recompute our state. I.e. if charm was in Waiting then that state mught be resolved when we call `_on_event()`
For this charm it is not critical, because it does not have many dependencies on other charms or events.

Reply:
Dropping `update-status` handling for now.